Use unique in `LoggerProvider` for storing loggers

yago-123 opened the pull request:
Replace internal map for keeping track of loggers in `LoggerProvider` with `unique` feature.

yago-123 commented on sdk/log/provider_test.go:
```diff
@@ -302,14 +302,14 @@ func TestLoggerProviderLogger(t *testing.T) {
+		assert.NotEqual(t, l0, l1)
+		assert.NotEqual(t, l0, l2)
+		assert.NotEqual(t, l1, l2)
 
 		l3, l4, l5 := p.Logger("l0"), p.Logger("l1"), p.Logger("l0", log.WithInstrumentationAttributes(attribute.String("foo", "bar")))
-		assert.Same(t, l0, l3)
-		assert.Same(t, l1, l4)
-		assert.Same(t, l2, l5)
+		assert.Equal(t, l0, l3)
+		assert.Equal(t, l1, l4)
+		assert.Equal(t, l2, l5)
```
This change is required because `Value()` retrieves a **shallow copy** of the inner `logger`:
```go
// Value returns a shallow copy of the T value that produced the Handle.
// Value is safe for concurrent use by multiple goroutines.
func (h Handle[T]) Value() T {
	return *h.value
}
```

Does this PR need benchmark tests?

pellared commented on sdk/log/provider.go:
```diff
@@ -118,22 +115,10 @@ func (p *LoggerProvider) Logger(name string, opts ...log.LoggerOption) log.Logge
-		p.loggers[scope] = l
-	}
+	l := newLogger(p, scope)
+	handle := unique.Make(l)
```
The `l` returned by `newLogger` is a new pointer. Unless I am not missing something, it will always create values which are not equal.

Moreover, we are not storing the handle anywhere. From https://go.dev/blog/unique:

> But Handle[T] also has a second purpose: as long as a Handle[T] exists for a value, the map will retain the canonical copy of the value. Once all Handle[T] values that map to a specific value are gone, the package marks that internal map entry as deletable, to be reclaimed in the near future. This sets a clear policy for when to remove entries from the map: when the canonical entries are no longer being used, then the garbage collector is free to clean them up.

yago-123 replied:
In any case, if this ends up in the garbage no worries, it's pretty interesting

pellared replied:
> Same was happening with the loggers:

Before the changes from this PR? If so then it is as suspected as this is the same pointer.

yago-123 replied:
Not sure if I get what you mean exactly. I added this test: 
```go
	t.Run("UniqueComparation", func(t *testing.T) {
		p := NewLoggerProvider()

		l0, l1 := p.Logger("l0"), p.Logger("l0")
		ul0, ul1 := unique.Make(l0), unique.Make(l1)

		// Should contain different pointer but same values (shallow copy)
		assert.Equal(t, ul0, ul1)
		assert.NotSame(t, l0, l1)

		l2 := p.Logger("l2")
		assert.NotEqual(t, l0, l2)
		assert.NotSame(t, l0, l2)
	})
```

yago-123 replied:
Yes, you are right, new pointer will be created because of shallow copy. That is why I also replaced `assert.Equal` for `assert.Same`. Because after calling unique a new pointer with a shallow copy will be created

pellared reviewed (commented):
Benchmakrs are a must-have to detemine if this even helps.

I also want to add that when I created https://github.com/open-telemetry/opentelemetry-go/issues/6352 as more of an "idea" - something that needs experiments. At quick glance I think we can intern `instrumentation.Scope` but not the whole logger. 

Possibly we could from Go 1.24 we could have weak pointers for `logger`.
I can imagine something like `loggers map[unique.Handle[instrumentation.Scope]]weak.Pointer[logger]` :sweat_smile:

pellared commented:
> there is not much doc tbh.

True. You can also check: https://victoriametrics.com/blog/go-unique-package-intern-string/

> Applying to Scope will only help comparing the values (still that would be acceptable to add).

TBH, the more I think about it I am not sure it will help at all. It is something that has to be measured (benchmarks).

yago-123 commented:
Closing this, apparently it only helps with amount of bytes allocated.

Old version vs. this branch version: 
```
benchmark                                                                old ns/op     new ns/op     delta
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     778           1903          +144.44%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     642           2757          +329.71%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     767           3790          +394.20%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     660           4317          +554.19%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     654           4951          +657.38%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            784           794           +1.17%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            785           965           +22.92%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            782           928           +18.65%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            779           1024          +31.45%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            819           1023          +24.86%

benchmark                                                                old allocs     new allocs     delta
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     3              6              +100.00%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     3              6              +100.00%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     3              6              +100.00%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     3              6              +100.00%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     3              6              +100.00%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            3              6              +100.00%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            3              6              +100.00%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            3              6              +100.00%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            3              6              +100.00%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            3              6              +100.00%

benchmark                                                                old bytes     new bytes     delta
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     420           220           -47.62%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     326           221           -32.21%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     403           224           -44.42%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     337           231           -31.45%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     338           233           -31.07%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            377           227           -39.79%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            380           231           -39.21%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            380           231           -39.21%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            380           234           -38.42%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            394           229           -41.88%
```


Old version vs. `map[unique.Handle[instrumentation.Scope]]` key (in this case not even a single metric improves): 
```
benchmark                                                                old ns/op     new ns/op     delta
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     778           1293          +66.09%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     642           1208          +88.28%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     767           1250          +62.99%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     660           1298          +96.70%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     654           1285          +96.57%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            784           1532          +95.33%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            785           1599          +103.72%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            782           1652          +111.23%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            779           1687          +116.56%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            819           1721          +110.06%

benchmark                                                                old allocs     new allocs     delta
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     3              7              +133.33%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     3              7              +133.33%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     3              7              +133.33%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     3              7              +133.33%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     3              7              +133.33%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            3              7              +133.33%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            3              7              +133.33%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            3              7              +133.33%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            3              7              +133.33%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            3              7              +133.33%

benchmark                                                                old bytes     new bytes     delta
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     420           437           +4.05%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     326           417           +27.91%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     403           423           +4.96%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     337           427           +26.71%
BenchmarkLoggerRetrieval/Create_different_loggers_single-threaded-12     338           406           +20.12%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            377           422           +11.94%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            380           421           +10.79%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            380           424           +11.58%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            380           424           +11.58%
BenchmarkLoggerRetrieval/Create_different_loggers_parallel-12            394           425           +7.87%

```

dmathieu commented:
Reducing the number of allocations is a pretty good thing.

dmathieu commented on sdk/log/logger_bench_test.go:
```diff
@@ -85,3 +87,25 @@ func newTestLogger(t testing.TB) log.Logger {
+	b.Run("Create different loggers parallel", func(b *testing.B) {
+		provider := NewLoggerProvider()
+
+		b.ReportAllocs()
```
```suggestion
		b.ReportAllocs()
		b.ResetTimer()
```

So creating the provider doesn't count towards the benchmark.

dmathieu commented on sdk/log/logger_bench_test.go:
```diff
@@ -85,3 +87,25 @@ func newTestLogger(t testing.TB) log.Logger {
+	b.Run("Create different loggers single-threaded", func(b *testing.B) {
+		provider := NewLoggerProvider()
+
+		b.ReportAllocs()
```
```suggestion
		b.ReportAllocs()
		b.ResetTimer()
```

pellared commented on sdk/log/logger_bench_test.go:
```diff
@@ -85,3 +87,25 @@ func newTestLogger(t testing.TB) log.Logger {
+
+		b.ReportAllocs()
+		for n := 0; n < b.N; n++ {
+			provider.Logger(fmt.Sprintf("test-%d", n))
```
None of the benchmarks checks if the interning helps when we retrieve **the same** logger multiple times

pellared replied:
Disregard my comment. It doesn't make sense

pellared commented:
> Reducing the number of allocations is a pretty good thing.

I think this is only because the PR removes the loggers map. We are returning new logger each time. We would need also a benchmark what happens if the caller retrieves "the same" logger multiple times.

pellared commented on sdk/log/logger_bench_test.go:
```diff
@@ -85,3 +87,25 @@ func newTestLogger(t testing.TB) log.Logger {
+
+		b.ReportAllocs()
+		for n := 0; n < b.N; n++ {
+			provider.Logger(fmt.Sprintf("test-%d", n))
```
Isn't strconv.Itoa more performant than using fmt?

yago-123 commented:
For the shake of science. Added prepopulated (I hope the impl. is correct) values old vs. new: 
``` 
benchmark                                                                                                       old ns/op     new ns/op     delta
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_0_prepopulated_values-12           827           715           -13.49%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_0_prepopulated_values-12                  759           890           +17.33%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_0_prepopulated_values#01-12        27.7          5619          +20170.56%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_0_prepopulated_values#01-12               64.0          932           +1356.79%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_1000_prepopulated_values-12        779           1017          +30.52%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_1000_prepopulated_values-12               739           950           +28.64%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_1000_prepopulated_values#01-12     35.6          5928          +16570.42%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_1000_prepopulated_values#01-12            68.8          937           +1260.35%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_5000_prepopulated_values-12        761           945           +24.30%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_5000_prepopulated_values-12               722           1115          +54.33%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_5000_prepopulated_values#01-12     35.8          6173          +17143.02%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_5000_prepopulated_values#01-12            68.6          1062          +1447.20%

benchmark                                                                                                       old allocs     new allocs     delta
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_0_prepopulated_values-12           2              5              +150.00%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_0_prepopulated_values-12                  2              5              +150.00%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_0_prepopulated_values#01-12        0              4              +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_0_prepopulated_values#01-12               0              4              +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_1000_prepopulated_values-12        2              5              +150.00%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_1000_prepopulated_values-12               2              5              +150.00%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_1000_prepopulated_values#01-12     0              4              +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_1000_prepopulated_values#01-12            0              4              +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_5000_prepopulated_values-12        2              5              +150.00%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_5000_prepopulated_values-12               2              5              +150.00%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_5000_prepopulated_values#01-12     0              4              +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_5000_prepopulated_values#01-12            0              4              +Inf%

benchmark                                                                                                       old bytes     new bytes     delta
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_0_prepopulated_values-12           379           217           -42.74%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_0_prepopulated_values-12                  368           220           -40.22%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_0_prepopulated_values#01-12        0             193           +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_0_prepopulated_values#01-12               0             189           +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_1000_prepopulated_values-12        373           218           -41.55%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_1000_prepopulated_values-12               356           223           -37.36%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_1000_prepopulated_values#01-12     0             200           +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_1000_prepopulated_values#01-12            0             192           +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_5000_prepopulated_values-12        369           217           -41.19%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_5000_prepopulated_values-12               347           212           -38.90%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_5000_prepopulated_values#01-12     0             198           +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_5000_prepopulated_values#01-12            0             187           +Inf%

```

yago-123 commented:
For the shake of science. Added the benchmarks for prepopulating the loggers with values and added bench for asking all the times the same different loggers in each iteration. A bit strange results: 
```
benchmark                                                                                                         old ns/op     new ns/op     delta
BenchmarkLoggerRetrieval/Retrieve_different_logger_each_time_single-threaded_with_0_prepopulated_values-12        811           757           -6.61%
BenchmarkLoggerRetrieval/Retrieve_different_logger_each_time_single-threaded_with_1000_prepopulated_values-12     740           854           +15.36%
BenchmarkLoggerRetrieval/Retrieve_different_logger_each_time_single-threaded_with_5000_prepopulated_values-12     757           919           +21.46%
BenchmarkLoggerRetrieval/Retrieve_different_logger_each_time_parallel_with_0_prepopulated_values-12               792           872           +10.06%
BenchmarkLoggerRetrieval/Retrieve_different_logger_each_time_parallel_with_1000_prepopulated_values-12            889           939           +5.55%
BenchmarkLoggerRetrieval/Retrieve_different_logger_each_time_parallel_with_5000_prepopulated_values-12            766           925           +20.70%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_0_prepopulated_values-12             28.1          5694          +20185.00%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_1000_prepopulated_values-12          35.8          5844          +16237.71%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_5000_prepopulated_values-12          35.9          5694          +15782.85%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_0_prepopulated_values-12                    63.0          879           +1294.54%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_1000_prepopulated_values-12                 70.0          883           +1161.72%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_5000_prepopulated_values-12                 69.0          900           +1204.03%

benchmark                                                                                                         old allocs     new allocs     delta
BenchmarkLoggerRetrieval/Retrieve_different_logger_each_time_single-threaded_with_0_prepopulated_values-12        2              5              +150.00%
BenchmarkLoggerRetrieval/Retrieve_different_logger_each_time_single-threaded_with_1000_prepopulated_values-12     2              5              +150.00%
BenchmarkLoggerRetrieval/Retrieve_different_logger_each_time_single-threaded_with_5000_prepopulated_values-12     2              5              +150.00%
BenchmarkLoggerRetrieval/Retrieve_different_logger_each_time_parallel_with_0_prepopulated_values-12               2              5              +150.00%
BenchmarkLoggerRetrieval/Retrieve_different_logger_each_time_parallel_with_1000_prepopulated_values-12            2              5              +150.00%
BenchmarkLoggerRetrieval/Retrieve_different_logger_each_time_parallel_with_5000_prepopulated_values-12            2              5              +150.00%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_0_prepopulated_values-12             0              4              +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_1000_prepopulated_values-12          0              4              +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_5000_prepopulated_values-12          0              4              +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_0_prepopulated_values-12                    0              4              +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_1000_prepopulated_values-12                 0              4              +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_5000_prepopulated_values-12                 0              4              +Inf%

benchmark                                                                                                         old bytes     new bytes     delta
BenchmarkLoggerRetrieval/Retrieve_different_logger_each_time_single-threaded_with_0_prepopulated_values-12        387           214           -44.70%
BenchmarkLoggerRetrieval/Retrieve_different_logger_each_time_single-threaded_with_1000_prepopulated_values-12     354           218           -38.42%
BenchmarkLoggerRetrieval/Retrieve_different_logger_each_time_single-threaded_with_5000_prepopulated_values-12     354           218           -38.42%
BenchmarkLoggerRetrieval/Retrieve_different_logger_each_time_parallel_with_0_prepopulated_values-12               361           223           -38.23%
BenchmarkLoggerRetrieval/Retrieve_different_logger_each_time_parallel_with_1000_prepopulated_values-12            374           222           -40.64%
BenchmarkLoggerRetrieval/Retrieve_different_logger_each_time_parallel_with_5000_prepopulated_values-12            361           220           -39.06%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_0_prepopulated_values-12             0             197           +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_1000_prepopulated_values-12          0             200           +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_single-threaded_with_5000_prepopulated_values-12          0             195           +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_0_prepopulated_values-12                    0             191           +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_1000_prepopulated_values-12                 0             200           +Inf%
BenchmarkLoggerRetrieval/Retrieve_same_logger_each_time_parallel_with_5000_prepopulated_values-12                 0             198           +Inf%
```

pellared commented:
> A bit strange results

Not really. Because:

> We are returning new logger each time.

We are also creating a new `unique.Handle` each time. See: https://github.com/open-telemetry/opentelemetry-go/pull/6431#discussion_r1993790195

yago-123 commented:
> We are also creating a new unique.Handle each time. See: https://github.com/open-telemetry/opentelemetry-go/pull/6431#discussion_r1993790195

Yes! I said strange for the 0 byte allocations in the old version. 

I guess we can close this PR, thank you both for your patience! I'm sorry it didn't perform as expected.

pellared commented:
> I guess we can close this PR, thank you both for your patience! I'm sorry it didn't perform as expected.

It is fine.   You can also blame me for creating the issue :wink: However, I hope we both learned something useful.

It seems that `unique` package does not fit our use case and `weak` should be used instead: https://go.dev/blog/cleanups-and-weak. However, it would require Go 1.24.
